corehq/apps/data_interfaces/models.py

```diff
@@ -1315,6 +1318,37 @@ def bulk_create_duplicate_relationships(cls, action, initial_case, duplicate_cas
         ]
         cls.potential_duplicates.through.objects.bulk_create(through_models)
 
+    @classmethod
+    def create_models(cls, case_ids, existing_case_ids, action):
+        case_duplicates = ([
+            cls(case_id=duplicate_case_id, action=action)
+            for duplicate_case_id in case_ids
+            if duplicate_case_id not in existing_case_ids
+        ])
+
+        for duplicate in case_duplicates:
+            case = CommCareCase.objects.get_case(duplicate.case_id)
+            duplicate.match_values = cls.case_and_action_to_hash(case, action)
+
+        return cls.objects.bulk_create(case_duplicates)
+
+    @classmethod
+    def case_and_action_to_hash(cls, case, action):
+        current_values = []
+        for prop in action.case_properties:
+            properties = case.resolve_case_property(prop)
+            current_values.extend([prop.value for prop in properties])
+
+        return cls.hash_arguments(*current_values)
+
+    @classmethod
+    def hash_arguments(cls, *args):
+        combined = hashlib.sha256()
+        for arg in args:
+            combined.update(str(arg).encode('utf8'))
```

Comment:
Combining `args` (property values) this way seems error prone. For example, imagine case A has values `['mans', 'laughter']` and case B has values `['man', 'slaughter']`. With this algorithm those two cases would appear to be duplicates even though they are not.

The best way I can think of to avoid that is to use CSV encoding:
```suggestion
        import csv
        class updater:
            def write(value):
                combined.update(value.encode('utf8'))

        combined = hashlib.sha256()
        writer = csv.writer(
            updater,
            delimiter="\t",
            quotechar=None,
            escapechar="|",
            quoting=csv.QUOTE_NONE,
        )
        writer.writerow(args)
```

Reply:
Thanks for catching this. I could have sworn I was joining them with a separator character before specifically to avoid this situation, but I must have left that out